packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -117,94 +131,118 @@ export default createRule<Options, MessageIds>({
-      duplicateConstituent: {
-        duplicated: TSESTree.TypeNode;
-        duplicatePrevious: TSESTree.TypeNode;
-      },
-      parentNode: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
-    ): void {
-      const beforeTokens = context.sourceCode.getTokensBefore(
-        duplicateConstituent.duplicated,
-        { filter: token => token.value === '|' || token.value === '&' },
-      );
-      const beforeUnionOrIntersectionToken =
-        beforeTokens[beforeTokens.length - 1];
-      const bracketBeforeTokens = context.sourceCode.getTokensBetween(
-        beforeUnionOrIntersectionToken,
-        duplicateConstituent.duplicated,
-      );
-      const bracketAfterTokens = context.sourceCode.getTokensAfter(
-        duplicateConstituent.duplicated,
-        { count: bracketBeforeTokens.length },
-      );
-      const reportLocation: TSESTree.SourceLocation = {
-        start: duplicateConstituent.duplicated.loc.start,
-        end:
-          bracketAfterTokens.length > 0
-            ? bracketAfterTokens[bracketAfterTokens.length - 1].loc.end
-            : duplicateConstituent.duplicated.loc.end,
-      };
-      context.report({
-        data: {
-          type:
-            parentNode.type === AST_NODE_TYPES.TSIntersectionType
-              ? 'Intersection'
-              : 'Union',
-          previous: context.sourceCode.getText(
-            duplicateConstituent.duplicatePrevious,
-          ),
-        },
-        messageId: 'duplicate',
-        node: duplicateConstituent.duplicated,
-        loc: reportLocation,
-        fix: fixer => {
-          return [
-            beforeUnionOrIntersectionToken,
-            ...bracketBeforeTokens,
-            duplicateConstituent.duplicated,
-            ...bracketAfterTokens,
-          ].map(token => fixer.remove(token));
-        },
-      });
-    }
+
     return {
-      ...(!ignoreIntersections && {
-        TSIntersectionType: checkDuplicate,
-      }),
+      ...(!ignoreIntersections && { TSIntersectionType: checkDuplicate }),
       ...(!ignoreUnions && {
-        TSUnionType: checkDuplicate,
+        TSUnionType: (node): void =>
+          checkDuplicate(node, (constituentNodeType, report) => {
+            const maybeTypeAnnotation = node.parent;
+            if (maybeTypeAnnotation.type === AST_NODE_TYPES.TSTypeAnnotation) {
+              const maybeIdentifier = maybeTypeAnnotation.parent;
+              if (
+                maybeIdentifier.type === AST_NODE_TYPES.Identifier &&
+                maybeIdentifier.optional
+              ) {
+                const maybeFunction = maybeIdentifier.parent;
+                const { type } = maybeFunction;
+                if (
+                  (type === AST_NODE_TYPES.ArrowFunctionExpression ||
+                    type === AST_NODE_TYPES.FunctionDeclaration ||
+                    type === AST_NODE_TYPES.FunctionExpression) &&
+                  maybeFunction.params.includes(maybeIdentifier) &&
```

Comment:
Shouldn't we also report on `TSDeclareFunction`, `TSEmptyBodyFunctionExpression`, `TSMethodSignature`, `TSFunctionType` and `TSAbstractMethodDefinition`?

```ts
declare function f(a?: string | undefined): void

declare class bb {
  f(a?: string | undefined): void
}

interface ee {
  f(a?: string | undefined): void
}

type fn = (a?: string | undefined) => void

abstract class cc {
  abstract f(a?: string | undefined): void
}
```

[playground](https://deploy-preview-9479--typescript-eslint.netlify.app/play/#ts=5.5.2&showAST=es&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVIAooB%2BALngGcMYtUBzeAH3jVEVpGAEpyA3HLMABQQ0JFgJxFCvABGs%2BAG8h8QiXJUa9Ji1RsO3PgOEBfEbQwgYiKGAQgEy1YiJlK1Wg2asQ7VJx54fkEhMyEMAE8ABwREfABeeFcND21vPV8DLnh4gD4g4xEoWU1bDHgpGTAwJRV4YtLMNTdNTx0fPwCjEJMgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false)

Reply:
Good catch! I added even more node types — I added all node types that had the `params` property I was using.
Also, note that `TSAbstractMethodDefinition` is not needed, because the `value` property of the `TSAbstractMethodDefinition` contains a `TSEmptyBodyFunctionExpression`.